### PR TITLE
Refactor: group DeepSeek V4 Flash LM head into per-group TP units

### PR DESCRIPTION
**Pull request by zhangqi-chen**
> [!IMPORTANT]
> Requires the matching pypto-serving change. Serving builds `lm_head_weight` and
> the LM-head argument list, so either side alone breaks the other: this branch
> expects one vocab shard per DP rank and no `num_logit_rows`, while `main`
> expects `TP_SIZE` shards and that argument.
>
> **Merge order:** this PR first, then the pypto-serving PR, which carries the
> serving-side change together with the `pypto-lib` submodule bump. The
> `serving-deepseek` check here therefore stays red until that PR lands — it
> builds pypto-serving `main`, which still has the old contract.

## Summary
- Cut the owner world into `--dp // --tp` independent groups: every card is both an owner and a TP rank, holding vocab shard `rank % TP_SIZE` and serving only the `TP_SIZE` owners of its own group, so every `peer` is `group_base + tp_rank`
- Collapse the four decoupled stage kernels into a single `lm_head` dispatch, with notify and wait sharing one `pl.at` scope so no window crosses orchestration as a new SSA version
- Take `lm_head_weight` as `[N_RANKS, VOCAB_PER_TP, D]` and read `weight[r]`: grouping puts every DP card on the LM head, and resident arguments are handed out per rank, so the shard must arrive replicated
- Derive the active logit rows from the `-1` sentinel in `logit_row_indices` and pass one row per card, dropping `num_tokens_per_owner`, `num_logit_rows` and `my_rank` from the kernel so its signature carries no DP extent
- Clamp the source row against the hidden extent so a stale index cannot build an illegal address
- Give `hidden_window` and `logits_window` their own window buffers so the LM head shares no storage with the MoE recv path, and restart the prefill barrier epoch at 1 now that it owns its counters
- Remove the legacy non-decoupled and decoupled LM head paths, the standalone workload proxy, and the `--grouped` / `--bench` / `--prefill-last-token` modes
- Rename the owner world from EP to DP, and drop the unused `LM_HEAD_TP_SIZE` config constant
- Rewire `decode_fwd` and `prefill_fwd` from four LM head stage loops to one `lm_head` call, take the decode `logits` row extent from `MAX_LOGIT_ROWS`, and assert `--ep % --tp == 0`
- Add `--active-ranks` to the prefill fixture to cover the single-request skew where most cards stay idle

### CODE REVIEW

File: models/deepseek/v4-flash/decode_fwd.py [hunk 1]
```diff
@@ -795,128 +788,43 @@ def l3_decode_fwd(
...
-        lm_head_logits_route_decoupled_worker(
-            tp_logits_shards[r], logits_window, logits_done,
-            r, LM_HEAD_COMM_EPOCH, device=r,
-        )
-
-    for r in pl.range(pld.world_size()):
-        logits_window = pld.window(recv_x_buf, [LM_HEAD_T_MAX, LM_HEAD_VOCAB], dtype=pl.FP32)
+        hidden_window = pld.window(recv_x_buf, [MAX_LOGIT_ROWS, D], dtype=pl.BF16)
+        hidden_done = pld.window(lm_head_hidden_done_buf, [LM_HEAD_TP_SIZE, 1], dtype=pl.INT32)
+        logits_window = pld.window(recv_x_buf, [MAX_LOGIT_ROWS, LM_HEAD_VOCAB], dtype=pl.FP32)
```
[the remark this answers is not in the document]
Comment (zhangqi-chen) [reply models/deepseek/v4-flash/decode_fwd.py]: Fixed in 6d741c99. `logits_window` now has its own `lm_head_logits_window_buf` in both drivers; only `hidden_window` still rides `recv_x_buf`, which stays safe because MoE and the LM head run sequentially. The race was real and cross-rank: `route_logits_shard` remote_stores into every group member's `logits_window`, and the only barrier before that point is "all peers published", not "all peers gathered", so a peer could still be reading those bytes as hidden rows.

File: models/deepseek/v4-flash/decode_fwd.py [hunk 2]
```diff
@@ -795,128 +788,43 @@ def l3_decode_fwd(
...
         logits_done = pld.window(lm_head_logits_done_buf, [LM_HEAD_TP_SIZE, 1], dtype=pl.INT32)
-        lm_head_finish_decoupled_worker(
-            logits[r], logits_window, logits_done,
-            r, LM_HEAD_COMM_EPOCH, device=r,
+        lm_head(
+            hidden_out[r], lm_head_weight[r % LM_HEAD_TP_SIZE], num_tokens_per_owner[r],
+            logit_row_indices[r], num_logit_rows[r], logits[r], hidden_window, hidden_done,
+            logits_window, logits_done, r // LM_HEAD_TP_SIZE * LM_HEAD_TP_SIZE,
+            r % LM_HEAD_TP_SIZE, LM_HEAD_COMM_EPOCH, device=r,
```
[the remark this answers is not in the document]
Comment (zhangqi-chen) [reply models/deepseek/v4-flash/decode_fwd.py]: Fixed in 6d741c99. Decode `logits` is now `[N_RANKS, MAX_LOGIT_ROWS, LM_HEAD_VOCAB]`, matching prefill and lm_head's output contract. The old `T` spelling only lined up because DECODE_BATCH*DECODE_SEQ happens to equal MAX_LOGIT_ROWS today; the emitted shape is unchanged.

File: models/deepseek/v4-flash/prefill_fwd.py [hunk 3]
```diff
@@ -845,74 +837,42 @@ def l3_prefill_fwd(
...
-        )
-
-    for r in pl.range(pld.world_size()):
-        logits_window = pld.window(
-            recv_x_buf, [LM_HEAD_T_MAX, LM_HEAD_VOCAB], dtype=pl.FP32,
-        )
+        hidden_window = pld.window(recv_x_buf, [MAX_LOGIT_ROWS, D], dtype=pl.BF16)
+        hidden_done = pld.window(arrived_buf, [LM_HEAD_TP_SIZE, 1], dtype=pl.INT32)
+        logits_window = pld.window(recv_x_buf, [MAX_LOGIT_ROWS, LM_HEAD_VOCAB], dtype=pl.FP32)
         logits_done = pld.window(combine_arrived_buf, [LM_HEAD_TP_SIZE, 1], dtype=pl.INT32)
```
[the remark this answers is not in the document]
Comment (zhangqi-chen) [reply models/deepseek/v4-flash/prefill_fwd.py]: Fixed in 6d741c99. Prefill now allocates `lm_head_hidden_done_buf` and `lm_head_logits_done_buf` instead of windowing `arrived_buf` / `combine_arrived_buf`, matching decode_fwd. The LM-head barrier no longer shares storage with the MoE epoch counters, which use different epoch semantics.
